File: web/src/app/core/api/notifications.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';

export type NotificationType = 'SLA_BREACH' | 'FOLLOW_UP_DUE';

export interface AppNotification {
  id: string;
  type: NotificationType;
  leadId: string | null;
  leadName: string | null;
  readAt: string | null;
  createdAt: string;
}

export function notificationText(n: AppNotification): string {
  const name = n.leadName ?? 'A lead';
  switch (n.type) {
    case 'SLA_BREACH':
      return `${name} is going cold — no contact within SLA`;
    case 'FOLLOW_UP_DUE':
      return `Follow-up due for ${name}`;
  }
}

@Injectable({ providedIn: 'root' })
export class NotificationsService {
  private readonly http = inject(HttpClient);

  list() {
    return this.http.get<AppNotification[]>('/api/notifications');
  }

  unreadCount() {
    return this.http.get<{ count: number }>('/api/notifications/unread-count');
  }

  markRead(id: string) {
    return this.http.post<void>(`/api/notifications/${id}/read`, {});
  }

  markAllRead() {
    return this.http.post<void>('/api/notifications/read-all', {});
  }
}

File: app/lib/notifications/notifications_service.dart
import '../core/api_client.dart';

class AppNotification {
  const AppNotification(this.json);

  final Map<String, dynamic> json;

  String get id => json['id'] as String;
  String get type => json['type'] as String;
  String? get leadId => json['leadId'] as String?;
  String? get leadName => json['leadName'] as String?;
  String? get readAt => json['readAt'] as String?;
  String get createdAt => json['createdAt'] as String;

  bool get isRead => readAt != null;

  String get text {
    final name = leadName ?? 'A lead';
    return switch (type) {
      'SLA_BREACH' => '$name is going cold — no contact within SLA',
      'FOLLOW_UP_DUE' => 'Follow-up due for $name',
      _ => name,
    };
  }
}

class NotificationsService {
  const NotificationsService._();

  static const NotificationsService instance = NotificationsService._();

  Future<List<AppNotification>> list() async {
    final data = await ApiClient.instance.get('/api/notifications') as List<dynamic>;
    return data
        .map((e) => AppNotification(e as Map<String, dynamic>))
        .toList(growable: false);
  }

  Future<int> unreadCount() async {
    final data = await ApiClient.instance.get('/api/notifications/unread-count')
        as Map<String, dynamic>;
    return (data['count'] as num).toInt();
  }

  Future<void> markRead(String id) async {
    await ApiClient.instance.post('/api/notifications/$id/read', {});
  }

  Future<void> markAllRead() async {
    await ApiClient.instance.post('/api/notifications/read-all', {});
  }
}
